frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseRepositoryAction.java

```diff
@@ -0,0 +1,292 @@
+    val presentation = anActionEvent.getPresentation();
+    if (!presentation.isEnabledAndVisible()) {
+      return;
+    }
+    presentation.setDescription(getNonHtmlString("action.GitMachete.TraverseRepositoryAction.description.text"));
+
+    val repositorySnapshot = getGitMacheteRepositorySnapshot(anActionEvent);
+    val branchLayout = repositorySnapshot != null ? repositorySnapshot.getBranchLayout() : null;
+
+    if (branchLayout == null || branchLayout.getRootEntries().isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription(getNonHtmlString("action.GitMachete.TraverseRepositoryAction.description.empty-layout"));
+      return;
+    }
+  }
+
+  @Override
+  @UIEffect
+  public void actionPerformed(AnActionEvent anActionEvent) {
+    val repository = getSelectedGitRepository(anActionEvent);
+    val branchLayout = getBranchLayout(anActionEvent);
+
+    if (branchLayout != null && branchLayout.getRootEntries().nonEmpty() && repository != null) {
+      boolean traverseApproved = true;
+      if (PropertiesComponent.getInstance().getBoolean(SHOW_TRAVERSE_APPROVAL, /* defaultValue */ true)) {
+
+        val traverseApprovalDialogBuilder = MessageDialogBuilder.okCancel(
+            getString("action.GitMachete.TraverseRepositoryAction.dialog.traverse-approval.title"),
+            getString("action.GitMachete.TraverseRepositoryAction.dialog.traverse-approval.text.HTML"));
+
+        traverseApprovalDialogBuilder
+            .yesText(getString("action.GitMachete.TraverseRepositoryAction.dialog.traverse-approval.ok-text"))
+            .icon(Messages.getQuestionIcon()).doNotAsk(new TraverseApprovalDialog());
+
+        traverseApproved = traverseApprovalDialogBuilder.ask(getProject(anActionEvent));
+      }
+
+      if (traverseApproved) {
+        var currentBranchName = getCurrentBranchNameIfManaged(anActionEvent);
+        var previousEntry = currentBranchName != null ? branchLayout.findPreviousEntry(currentBranchName) : null;
+        while (previousEntry != null) {
+          val checkoutPreviousAction = ActionManager.getInstance()
+              .getAction(actionIdFormatString.format(CheckoutPreviousAction.class.getSimpleName()));
+          checkoutPreviousAction.actionPerformed(anActionEvent);
+          currentBranchName = previousEntry.getName();
+          previousEntry = branchLayout.findPreviousEntry(currentBranchName);
+        }
+
+        if (currentBranchName != null) {
+          traverse(currentBranchName, anActionEvent);
+        }
+
+      }
+    }
+  }
+
+  @UIEffect
+  private void traverse(String branchName, AnActionEvent anActionEvent) {
+    val gitMacheteBranch = getManagedBranchByName(anActionEvent, branchName);
+
+    if (gitMacheteBranch != null) {
+      val remoteTrackingBranch = gitMacheteBranch.getRemoteTrackingBranch();
+      if (remoteTrackingBranch != null) {
+        syncBranchToRemote(remoteTrackingBranch, gitMacheteBranch, anActionEvent);
```

Comment:
Hmm so there can be two syncs to remote for the same branch (first here, then in line 133)? I'd rather suggest sync to remote only once, as in the CLI

Reply:
the initial rationale was that the sync to parent, for example by rebase, can cause loss of sync to remote. not necessary?

Reply:
I'd keep sync to remote for only after sync to parent, BUT I'd skip the sync to parent in case:
* branch is behind its remote
* branch is diverged from & older then its remote

Rel: https://github.com/VirtusLab/git-machete/issues/292

Reply:
Super. Got it.